### PR TITLE
feat: support dynamically changing qos for EIP

### LINKED ISSUE #2633: More QoS features supported in NATGW
### Feature request

Trying to implement more QoS features supported in NATGW.

### Use case

The following QoS features will be supported:
* Bandwidth rate limit for NATGW net1 NIC
* Bandwidth rate limit for NATGW net1 EIP (already supported in #2496)
* Bandwidth rate limit for external network traffic on NATGW net1

### PR DESCRIPTION
**Pull request by shane965**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

### Which issue(s) this PR fixes:
Fixes #2633

### <samp>🤖 Generated by Copilot at 54e9570</samp>

This pull request adds support for multiple QoS policy rules with different actions and match criteria, and improves the QoS policy deep copy, validation, and reconciliation logic. It also updates the QoSPolicy CRD and the QoS policy functions for EIPs to use the new QoS policy spec fields and types.

### <samp>🤖 Generated by Copilot at 54e9570</samp>

> _To manage the QoS policy rules_
> _We added some types and some tools_
> _With `action` and `direction`_
> _And `bandwidthLimit` selection_
> _We can handle more cases and use cases_

### <samp>🤖 Generated by Copilot at 54e9570</samp>

*  Add new types and constants for QoS policy rules and modify QoSPolicySpec and QoSPolicyBandwidthLimitRule types to support multiple rules with different parameters and actions in `pkg/apis/kubeovn/v1/types.go` ([link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR4-R5), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR44-R62), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL1269-R1292), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL1295-R1342))
* Modify QoSPolicy, QoSPolicySpec, and QoSPolicyStatus types to deep copy the new fields in `pkg/apis/kubeovn/v1/zz_generated.deepcopy.go` ([link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-06f8e5bc84406eb8a89c72c345fc4ff5f144e480103ad55d681f7567f0f85515L1267-R1267), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-06f8e5bc84406eb8a89c72c345fc4ff5f144e480103ad55d681f7567f0f85515L1360-R1370), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-06f8e5bc84406eb8a89c72c345fc4ff5f144e480103ad55d681f7567f0f85515L1377-R1397))
* Modify QoSPolicy CRD schema and additionalPrinterColumns to reflect the new fields in `yamls/crd.yaml` ([link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5L1967-R1971), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5L1980-R2004), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5L2007-R2049))
* Add a function to compare QoS policy rules and return the added, deleted, and updated rules in `pkg/controller/qos_policy.go` ([link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662R32-R44))
* Modify QoS policy controller methods to handle different cases of QoS policy update, validate QoS policy rule names, and reconcile QoS policy rules for related EIPs in `pkg/controller/qos_policy.go` ([link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662L44-R62), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662L180-R200), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662L186-R210), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662L194-R219), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662L203-R230), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662R275-R350), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662L259-R363), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-8aa74956ce96fc812f32fcad39869a83b2fe26b5cf48958c78e41c29a56ca662R385-R427))
* Modify EIP QoS controller methods to apply or delete QoS policy rules based on the new fields and parameters in `pkg/controller/vpc_nat_gw_eip.go` ([link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L495-R540), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L527-R558), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L549-R571), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L565-R581), [link](https://github.com/kubeovn/kube-ovn/pull/2671/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L575-R593))

### CODE REVIEW

**Comment by shane965** [comment]
could you please review this pr? @bobz965  @hongzhen-ma

File: dist/images/vpcnatgateway/nat-gateway.sh [hunk 1]
```diff
@@ -276,17 +279,19 @@ function eip_ingress_qos_add() {
     do
```
Comment (bobz965) [inline dist/images/vpcnatgateway/nat-gateway.sh:279]: # ingress: 
# external --> net1 --> qos -->
# dst ip is iptables eip on net1

File: dist/images/vpcnatgateway/nat-gateway.sh [hunk 2]
```diff
@@ -295,16 +300,18 @@ function eip_egress_qos_add() {
...
```
Comment (bobz965) [inline dist/images/vpcnatgateway/nat-gateway.sh:300]:
``` bash
# ingress: 
# net1-->qos --> external
# src ip is iptables eip on net1

```

File: pkg/apis/kubeovn/v1/types.go [hunk 3]
```diff
@@ -1292,14 +1315,31 @@ type QoSPolicyCondition struct {
...
+	return fmt.Sprintf("%s", resultNames)
 }
 
 type QoSPolicyStatus struct {
-	// +optional
-	// +patchStrategy=merge
-	BandwidthLimitRule QoSPolicyBandwidthLimitRule `json:"bandwidthLimitRule" patchStrategy:"merge"`
+	BandwidthLimitRules QoSPolicyBandwidthLimitRules `json:"bandwidthLimitRules" patchStrategy:"merge"`
+	Shared              bool                         `json:"shared" patchStrategy:"merge"`
+	ActionType          QoSPolicyRulesType           `json:"actionType" patchStrategy:"merge"`
```
Comment (bobz965) [inline pkg/apis/kubeovn/v1/types.go:1341]: 这个对应到eip，感觉更偏向于资源类型，而非action

File: pkg/controller/qos_policy.go [hunk 4]
```diff
@@ -27,6 +29,19 @@ func (c *Controller) enqueueAddQoSPolicy(obj interface{}) {
...
+	if len(old) != len(new) {
+		return false
+	}
+	sort.Slice(old, func(i, j int) bool {
+		return old[i].Name < old[j].Name
+	})
+	sort.Slice(new, func(i, j int) bool {
+		return new[i].Name < new[j].Name
+	})
+	return reflect.DeepEqual(old, new)
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:42]: reflect.DeepEqual(old, new) 本身是否就足够了？

File: pkg/controller/qos_policy.go [hunk 5]
```diff
@@ -285,6 +382,49 @@ func (c *Controller) handleUpdateQoSPolicy(key string) error {
 		klog.Errorf("failed to handle add finalizer for qos, %v", err)
 		return err
 	}
+
+	if cachedQos.Status.Shared != cachedQos.Spec.Shared ||
+		cachedQos.Status.ActionType != cachedQos.Spec.ActionType {
+		err := fmt.Errorf("not support qos %s change shared,type ", key)
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:388]: err := fmt.Errorf("not support qos %s change type shared", key)

File: pkg/controller/vpc_nat_gw_eip.go [hunk 6]
```diff
@@ -492,29 +492,49 @@ func (c *Controller) deleteEipInPod(dp, v4Cidr string) error {
...
-			klog.Errorf("failed to set ingress eip '%s' qos in pod, %v", eip.Name, err)
-			return err
-		}
-	}
-
-	// set egress qos
-	if egressRate != "" {
-		if err = c.addEipQoSInPod(eip.Spec.NatGwDp, v4ip, util.QoSDirectionEgress, egressRate); err != nil {
-			klog.Errorf("failed to set egress eip '%s' qos in pod, %v", eip.Name, err)
+func (c *Controller) delEIPBandtithLimitRules(eip *kubeovnv1.IptablesEIP, v4ip string, rules kubeovnv1.QoSPolicyBandwidthLimitRules) error {
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gw_eip.go:532]: if !qosPolicy.Spec.Shared { 则不执行

File: pkg/controller/vpc_nat_gw_eip.go [hunk 7]
```diff
@@ -492,29 +492,49 @@ func (c *Controller) deleteEipInPod(dp, v4Cidr string) error {
...
+		}
+	}
+	return nil
+}
+
 // add tc rule for eip in nat gw pod
 func (c *Controller) addEipQoS(eip *kubeovnv1.IptablesEIP, v4ip string) error {
 	var err error
 	qosPolicy, err := c.config.KubeOvnClient.KubeovnV1().QoSPolicies().Get(context.Background(), eip.Spec.QoSPolicy, metav1.GetOptions{})
+	if !qosPolicy.Spec.Shared {
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gw_eip.go:510]: if !qosPolicy.Status.Shared 更准确，spec用户直接改就生效了

File: dist/images/vpcnatgateway/nat-gateway.sh (hunk 1 above)
Comment (bobz965) [reply dist/images/vpcnatgateway/nat-gateway.sh:279]: 我想补充这些注释

File: pkg/controller/qos_policy.go [hunk 8]
```diff
@@ -245,6 +275,68 @@ func (c *Controller) handleDelQoSPoliciesFinalizer(key string) error {
...
+	// Create a map of old structs indexed by name for efficient lookup
+	oldMap := make(map[string]*kubeovnv1.QoSPolicyBandwidthLimitRule)
+	for _, s := range oldList {
+		oldMap[s.Name] = s
+	}
+
+	// Loop through new structs and compare with old structs
+	for _, s := range newList {
+		if old, ok := oldMap[s.Name]; !ok {
+			// New struct
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:292]: // add rule

File: pkg/controller/qos_policy.go [hunk 9]
```diff
@@ -245,6 +275,68 @@ func (c *Controller) handleDelQoSPoliciesFinalizer(key string) error {
...
+			added = append(added, s)
+		} else if !reflect.DeepEqual(old, s) {
+			// Updated struct
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:295]: update changed rule

File: pkg/controller/qos_policy.go [hunk 10]
```diff
@@ -245,6 +275,68 @@ func (c *Controller) handleDelQoSPoliciesFinalizer(key string) error {
...
+			updated = append(updated, s)
+		}
+		// Remove old struct from map
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:298]:
```suggestion
		// rule not exist
```

File: pkg/controller/qos_policy.go [hunk 11]
```diff
@@ -245,6 +275,68 @@ func (c *Controller) handleDelQoSPoliciesFinalizer(key string) error {
...
+		if err = c.addOrUpdateEIPBandtithLimitRules(eip, eip.Status.IP, added); err != nil {
+			return err
+		}
+	}
+	if len(updated) > 0 {
+		if err = c.addOrUpdateEIPBandtithLimitRules(eip, eip.Status.IP, updated); err != nil {
+			return err
+		}
+	}
+
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:332]: 根据命名， add or update 合并一次性添加会更好些

File: pkg/controller/qos_policy.go [hunk 12]
```diff
@@ -245,6 +275,68 @@ func (c *Controller) handleDelQoSPoliciesFinalizer(key string) error {
...
+	return nil
+}
+
+func (c *Controller) validateQosPolicy(qosPolicy *kubeovnv1.QoSPolicy) error {
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:336]: 空函数

File: pkg/controller/qos_policy.go [hunk 13]
```diff
@@ -285,6 +371,65 @@ func (c *Controller) handleUpdateQoSPolicy(key string) error {
...
+		} else {
+			if cachedQos.Status.BindingType == kubeovnv1.QoSBindingTypeEIP {
+				// filter to eip
+				eips, err := c.config.KubeOvnClient.KubeovnV1().IptablesEIPs().List(context.Background(),
+					metav1.ListOptions{LabelSelector: fields.OneTermEqualSelector(util.QoSLabel, key).String()})
+				if err != nil {
+					klog.Errorf("failed to get eip list, %v", err)
+					return err
+				}
+				if len(eips.Items) == 0 {
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:406]: len(eips.Items) --> len(eips) 应该就可以

File: pkg/controller/qos_policy.go [hunk 14]
```diff
@@ -285,6 +371,65 @@ func (c *Controller) handleUpdateQoSPolicy(key string) error {
...
+					// not thing to do
+				} else if len(eips.Items) == 1 {
+					eip := &eips.Items[0]
+					if err = c.reconcileEIPBandtithLimitRules(eip, added, deleted, updated); err != nil {
+						return nil
+					}
+				} else {
+					err := fmt.Errorf("not support qos %s change rule, related eip more than one", key)
+					klog.Error(err)
+					// update status to failed?
```
Comment (bobz965) [inline pkg/controller/qos_policy.go:416]: 这个后续如果有设计condition，可以记录下状态，类似subnet那里，可以改为todo提示

File: pkg/controller/vpc_nat_gw_eip.go [hunk 15]
```diff
@@ -492,29 +492,49 @@ func (c *Controller) deleteEipInPod(dp, v4Cidr string) error {
...
 	var err error
 	qosPolicy, err := c.config.KubeOvnClient.KubeovnV1().QoSPolicies().Get(context.Background(), eip.Spec.QoSPolicy, metav1.GetOptions{})
+	if !qosPolicy.Status.Shared {
+		eips, err := c.config.KubeOvnClient.KubeovnV1().IptablesEIPs().List(context.Background(),
+			metav1.ListOptions{LabelSelector: fields.OneTermEqualSelector(util.QoSLabel, qosPolicy.Name).String()})
+		if err != nil {
+			klog.Errorf("failed to get eip list, %v", err)
+			return err
+		}
+		if len(eips.Items) != 0 {
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gw_eip.go:517]: 同上

File: pkg/controller/vpc_nat_gw_eip.go [hunk 16]
```diff
@@ -492,29 +492,49 @@ func (c *Controller) deleteEipInPod(dp, v4Cidr string) error {
...
+			if eips.Items[0].Name != eip.Name {
```
Comment (bobz965) [inline pkg/controller/vpc_nat_gw_eip.go:518]: eips[0].Name 是否可以？

File: pkg/controller/qos_policy.go (hunk 13 above)
Comment (shane965) [reply pkg/controller/qos_policy.go:406]: 这块应该不行eips是一个这样的数据结构：
```go
type IptablesEIPList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata"`

	Items []IptablesEIP `json:"items"`
}
```

File: pkg/controller/qos_policy.go (hunk 14 above)
Comment (shane965) [reply pkg/controller/qos_policy.go:416]: 这段我去掉吧，是我写伪代码时留下的

File: pkg/controller/vpc_nat_gw_eip.go (hunk 16 above)
Comment (shane965) [reply pkg/controller/vpc_nat_gw_eip.go:518]: 不行eips是个结构体

File: pkg/controller/qos_policy.go (hunk 12 above)
Comment (shane965) [reply pkg/controller/qos_policy.go:336]: 我预留的，后面的NATGW QoS 的 PR会在这里加校验逻辑

File: pkg/controller/qos_policy.go (hunk 11 above)
Comment (shane965) [reply pkg/controller/qos_policy.go:332]: 感觉这样会清晰一点，如果后续要用tc replace只要调整updated这块的`addOrUpdateEIPBandtithLimitRules`就行了

File: pkg/controller/vpc_nat_gw_eip.go (hunk 16 above)
Comment (shane965) [reply pkg/controller/vpc_nat_gw_eip.go:518]: 这块查询考虑做以下调整已提高性能，后面我单独提pr一起改下这部分 @bobz965 @hongzhen-ma 
IptablesEIPs().List() -> iptablesEipsLister.List()

**Comment by oilbeater** [comment]
The yamls in chart should also be updated

File: pkg/controller/qos_policy.go [hunk 17]
```diff
@@ -285,6 +371,64 @@ func (c *Controller) handleUpdateQoSPolicy(key string) error {
...
+				if err != nil {
+					klog.Errorf("failed to get eip list, %v", err)
+					return err
+				}
+				if len(eips.Items) == 0 {
+					// not thing to do
+				} else if len(eips.Items) == 1 {
+					eip := &eips.Items[0]
+					if err = c.reconcileEIPBandtithLimitRules(eip, added, deleted, updated); err != nil {
+						return nil
```
Comment (hongzhen-ma) [inline pkg/controller/qos_policy.go:411]: return err。 并且记录log

**Review by hongzhen-ma — APPROVED** [review]

**Pull request merged**